### PR TITLE
[SPARK-49022][CONNECT][SQL][FOLLOW-UP] Parse unresolved identifier to keep the behavior same

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR is a followup of https://github.com/apache/spark/pull/47688 that keeps `Column.toString` as the same before.

### Why are the changes needed?

To keep the same behaviour with Spark Classic and Connect.

### Does this PR introduce _any_ user-facing change?

No, the main change has not been released out yet.

### How was this patch tested?

Will be added separately. I manually tested:

```scala
import org.apache.spark.sql.functions.col
val name = "with`!@#$%dot".replace("`", "``")
col(s"`${name}`").toString.equals("with`!@#$%dot")
```

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @hvanhovell

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -413,7 +414,11 @@ class Dataset[T] private[sql] (
 
   /** @inheritdoc */
   def metadataColumn(colName: String): Column = {
-    Column(UnresolvedAttribute(colName, getPlanId, isMetadataColumn = true))
+    Column(
+      UnresolvedAttribute(
+        AttributeNameParser.parseAttributeName(colName),
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:419]: Can you move this into an `apply` method? It seems easy enough to forget.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/internal/columnNodeSupport.scala [hunk 2]
```diff
@@ -52,9 +52,9 @@ object ColumnNodeToProtoConverter extends (ColumnNode => proto.Expression) {
       case Literal(value, Some(dataType), _) =>
         builder.setLiteral(toLiteralProtoBuilder(value, dataType))
 
-      case UnresolvedAttribute(unparsedIdentifier, planId, isMetadataColumn, _) =>
+      case u @ UnresolvedAttribute(unparsedIdentifier, planId, isMetadataColumn, _) =>
         val b = builder.getUnresolvedAttributeBuilder
-          .setUnparsedIdentifier(unparsedIdentifier)
+          .setUnparsedIdentifier(u.sql)
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/internal/columnNodeSupport.scala:57]: Can we use something a bit more descriptive here? For example `escapedName`?

File: sql/core/src/main/scala/org/apache/spark/sql/internal/columnNodeSupport.scala [hunk 3]
```diff
@@ -74,7 +74,7 @@ private[sql] trait ColumnNodeToExpressionConverter extends (ColumnNode => Expres
           analysis.UnresolvedRegex(columnNameRegex, Some(nameParts), conf.caseSensitiveAnalysis)
 
         case UnresolvedRegex(unparsedIdentifier, planId, _) =>
-          convertUnresolvedAttribute(unparsedIdentifier, planId, isMetadataColumn = false)
+          convertUnresolvedRegex(unparsedIdentifier, planId)
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/columnNodeSupport.scala:77]: This seems a bit inconsistent with unresolved attributes. Is this observable?

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/columnNodeSupport.scala:77]: I think it's fine in a way that it's consistent with catalyst ..

**Comment by HyukjinKwon** [comment]
Let me merge this @hvanhovell please let me know if there are some more things to change, I will make another PR.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**